packages/medusa/src/instrumentation/index.ts

```diff
@@ -82,16 +89,28 @@ export function instrumentHttpLayer() {
 
     return async (req, res, next) => {
       if (shouldExcludeResource(req.originalUrl)) {
-        return await handler(req, res, next)
+        return handler(req, res, next)
       }
 
       await HTTPTracer.trace(traceName, async (span) => {
-        try {
-          await handler(req, res, next)
-        } finally {
-          span.end()
-        }
+        return new Promise<void>((resolve, reject) => {
+          handler(req, res, (error?: any) => {
+            if (error) {
+              span.setStatus({
+                code: SpanStatusCode.ERROR,
+                message: error.message || "Failed",
+              })
+              span.end()
+              reject(error)
+            } else {
+              span.end()
+              resolve()
+            }
+          })
+        })
       })
+        .catch(next)
+        .then(next)
```

Comment:
Would it be more clear with this?

```ts
const next_ = (err?: any) => {
  if (err) {
    span.setStatus({
      code: SpanStatusCode.ERROR,
      message: error.message || "Failed",
    })
  }
  span.end()
  next(err)
}

return handler(req, res, next_)
```

Reply:
Yeah. Can move it out.

Reply:
But also is the nested promise required? Without it it should work since next_ will be called at the right time always by the handler

Reply:
Yeah, the promise is required. I can explain this over a call. Its because of some surprising behavior from OpenTelemetry